### PR TITLE
Improve encryption allocations

### LINKED ISSUE #552: Crypto performance improvements
1. `AsnKeyParser` can be made a `ref struct` and can operate on `ReadOnlySpan<byte>`. Needs benchmarking to evaluate performance. This will require taking a dependency on `System.Memory`.

2. `SymmetricEncryptWithHMACIV` doesn't need a separate array to hold random values. We can fill them directly into the end of the `iv` and read them from there.

3. `SymmetricDecrypt` doesn't need to copy the entire ciphertext into a new array. We can use the `MemoryStream` constructor that accepts an array, offset and length.

### PR DESCRIPTION
**Pull request by xPaw**
Supercedes #1378 (based on this pr)
Supercedes #1318
Supercedes #683
Fixes #552

This removes all public methods from `CryptoHelper` except for `SymmetricDecrypt` which is used by NHA and some other projects that use it for decrypting depots manifests/filenames.

I deleted `NetFilterEncryption` (non HMAC version) which does not appear to be used by Steam (I looked at node-steam-user code, they always skip the first 16 bytes in encrypted packets, so it must always be present).

I also changed envelope send to use arraypool instead of temporarily allocating a byte array for the encrypted message.

### CODE REVIEW

File: SteamKit2/SteamKit2/Networking/Steam3/WebSocketContext.cs [hunk 1]
```diff
@@ -88,9 +88,8 @@ async Task RunCore(TimeSpan connectionTimeout, CancellationToken cancellationTok
                 }
             }
 
-            public async Task SendAsync(byte[] data)
+            public async Task SendAsync(ArraySegment<byte> segment)
             {
-                var segment = new ArraySegment<byte>(data, 0, data.Length);
                 try
                 {
                     await socket.SendAsync(segment, WebSocketMessageType.Binary, true, cts.Token).ConfigureAwait(false);
```
Comment (xPaw) [inline SteamKit2/SteamKit2/Networking/Steam3/WebSocketContext.cs:95]: This api does not have a Span<> overload forcing the use of ArraySegment everywhere, which I don't like. Any ideas?

File: SteamKit2/SteamKit2/Networking/Steam3/UdpConnection.cs [hunk 2]
```diff
@@ -175,13 +175,13 @@ public void Disconnect( bool userInitiated )
...
-        /// <param name="data">The data to send to the server</param>
-        public void Send( byte[] data )
+        /// <param name="segment">The data to send to the server</param>
+        public void Send( ArraySegment<byte> segment )
         {
             if ( state != (int)State.Connected )
                 return;
 
-            using var ms = new MemoryStream( data );
+            using var ms = new MemoryStream( segment.ToArray() ); // TODO: Remove memorystream
```
Comment (xPaw) [inline SteamKit2/SteamKit2/Networking/Steam3/UdpConnection.cs:184]: TODO

File: SteamKit2/SteamKit2/Networking/Steam3/EnvelopeEncryptedConnection.cs [hunk 3]
```diff
@@ -47,14 +48,26 @@ public void Disconnect( bool userInitiated )
...
+                inner.Send( data );
+                return;
             }
 
-            inner.Send( data );
+            var buffer = ArrayPool<byte>.Shared.Rent( encryption!.CalculateMaxEncryptedDataLength( data.Length ) );
+            try
+            {
+                var length = encryption!.ProcessOutgoing( data.Span, buffer );
+                var segment = new Memory<byte>( buffer, 0, length );
```
Comment (yaakov-h) [inline SteamKit2/SteamKit2/Networking/Steam3/EnvelopeEncryptedConnection.cs:63]:
```suggestion
                var segment = buffer.AsMemory();
```

Comment (xPaw) [reply SteamKit2/SteamKit2/Networking/Steam3/EnvelopeEncryptedConnection.cs:63]: Need to use the length.

**Comment by xPaw** [comment]
Related reading: https://github.com/dotnet/runtime/issues/55601

**Pull request merged**